test/1.4/misc/size_stats.dml

```diff
@@ -0,0 +1,101 @@
+
+  First, the baseline function is a method body that generates about 10*64*20 =
+  12k bytes of code (-1 expands to a ~20 byte uint64 literal, and we output 10
+  variables with 64 * -1 as initializers).
+
+  Then two typical code explosion patterns are exercised: one is to instantiate
+  many non-shared implementations of a method, and one is to expand a large
+  body using #foreach or #select.
+
+  SIMICS-20398
+*/
+param bloat_factor = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
+
+template a { group x is b; group y is b; group z is b; group w is b; }
+template b {
+    shared method m1() -> (int) {
+        local int ret;
+        #foreach _ in (bloat_factor) {
+            local uint64 i[64] = {
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                };
+            ret += i[100];
+        }
+        return ret;
+    }
+
+    method m2() -> (int) {
+        local int ret;
+        #foreach _ in (bloat_factor) {
+            local uint64 i[64] = {
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                };
+            ret += i[100];
+        }
+        return ret;
+    }
+}
+
+template t {
+    method m3() -> (int) {
+        local int ret;
+        foreach b in (each b in (dev)) {
+            #foreach _ in (bloat_factor) {
+                local uint64 i[64] = {
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                };
+                ret += i[100];
+                b.m1();
+            }
+        }
+        return ret;
+    }
+
+    method m4() -> (int) {
+        local int ret;
+        #foreach b in ([x, y, z, w]) {
+            #foreach _ in (bloat_factor) {
+                local uint64 i[64] = {
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+                };
+                ret += i[100];
+                b.m2();
+            }
+        }
+        return ret;
+    }
+}
+
+is a;
+is t;
+
+method init() {
+    m3();
+    m4();
+}
```

Comment:
This test looks fine cept the `i[100]`:s. I'd be very careful to avoid DCE from mucking things up, but from a casual inspection nothing looks dead.

Reply:
Right, I did some redundant asserts in the test to validate that things aren't messed up by DCEs.